pkg/cloudprovider/aws/instance.go

```diff
@@ -171,8 +171,15 @@ func (p *InstanceProvider) getLaunchTemplateConfigs(ctx context.Context, constra
 
 func (p *InstanceProvider) getOverrides(instanceTypeOptions []cloudprovider.InstanceType, subnets []*ec2.Subnet, capacityType string) []*ec2.FleetLaunchTemplateOverridesRequest {
 	var overrides []*ec2.FleetLaunchTemplateOverridesRequest
+	zonesUsed := sets.String{}
 	for i, instanceType := range instanceTypeOptions {
-		for zone := range instanceType.Zones() {
+		for _, offering := range instanceType.Offerings() {
+			zone := offering.Zone
+			// we can't assume that all zones will be available for all capacity types, hence this check
+			// (we shouldn't ever have duplicate zones, just being paranoid to create a well-formed set of overrides)
```

Comment:
This type of check is like wearing kevlar in your own living room in case you accidentally shoot yourself. There's no customer exposed surface here and Offerings() is a set (right?). IMO we should skip the zone uniqueness check.

Reply:
Offerings() is _not_ a set, it's a slice, so I fundamentally/philosophically disagree with you here. A duplicate is very much _possible_. A bug could easily be introduced where we have duplicate offerings. Because there's no constructor like Java, it's not like I can validate the initialization of the InstanceType struct (correct me if I'm wrong here), so there's nothing to maintain the invariant that the Offerings are all unique, unless I invent a set type specifically for Offering, which seems like way more work than adding 2 lines of code here.

The downside if I'm wrong and we never have a duplicate is we have an extra set that has a handful of elements in it and an extra conditional check. The downside if you're wrong is that users will fail to scale out because of invalid LT overrides.

Reply:
I hear ya'. I don't think there's going to be a great solution to this type of thing until Go adds generics to the language, at which point it would be pretty easy to make Offerings() return a set. FWIW, I think that the "house style" of a lot of Go code (in Kubernetes land) is to assume validation has occurred elsewhere and not keep checking invariants. Personally, I would suggest you just make the call here whether to keep or not, although I would probably remove the check myself.

Reply:
Ellis and I came up with a good compromise - I'm going to move the validation to the instance type provider. Update coming momentarily.

Team work makes the dream work :rainbow: